### PR TITLE
add NewClientFromPublishSettingsFile

### PR DESCRIPTION
**Pull request by waitingkuo**
#57 #58 

Add a function NewClientFromPublishSettingsFile() to create a client from the publishsettings file

Signed-off-by: Wei-Ting Kuo <waitingkuo0527@gmail.com>

### CODE REVIEW

File: management/client.go [hunk 1]
```diff
@@ -54,6 +54,21 @@ func NewClientFromConfig(subscriptionID string, managementCert []byte, config Cl
 	return makeClient(subscriptionID, managementCert, config.ManagementURL)
 }
 
+// NewClientFromPublishSettingsFile creates a new Client using a publishsettings file
+func NewClientFromPublishSettingsFile(filePath string) (Client, error) {
```
Comment (ahmetalpbalkan) [inline management/client.go]: We need another constructor that take publish settings file and config (like NewClientFromConfig)

How about these two:
- NewClientFromPublishSettings
- NewClientFromPublishSettingsAndConfig

see the existing constructors above.

File: management/client.go [hunk 2]
```diff
@@ -54,6 +54,21 @@ func NewClientFromConfig(subscriptionID string, managementCert []byte, config Cl
...
+
+	client := Client{
+		managementURL: defaultAzureManagementURL,
+	}
+
+	err := client.importPublishSettingsFile(filePath)
+	if err != nil {
```
Comment (ahmetalpbalkan) [inline management/client.go]: no need for this err check, just `return client, err` below.

**Comment by waitingkuo** [comment]
Updated

**Comment by ahmetalpbalkan** [comment]
@waitingkuo please update the other comment as well. also please squash your commits into one before pushing. Thanks!

**Comment by ahmetalpbalkan** [comment]
@waitingkuo ping, can you update? otherwise #65 also addresses this problem.

**Comment by waitingkuo** [comment]
Hi @ahmetalpbalkan 

`publishSettings` is private
https://github.com/MSOpenTech/azure-sdk-for-go/blob/master/management/publishSettings.go#L112-L117

Users cannot construct the client by `publishSettings`. Need I make it public?

**Comment by ahmetalpbalkan** [comment]
@waitingkuo the only thing I'm asking for: currently what you have is:

    func NewClientFromPublishSettingsFile(filePath string) (Client, error)

and I'm saying let's have these two instead:

    func NewClientFromPublishSettingsFile(publishSettings string) (Client, error)
    func NewClientFromPublishSettingsFileAndConfig(publishSettings string, configConfig) (Client, error)

You don't need to accept `publishSettings` type as parameter. It's for the parsed publishSettings file and should not be used as input.

**Comment by waitingkuo** [comment]
Sorry for the misunderstanding, just updated

File: management/client.go [hunk 3]
```diff
@@ -54,6 +54,30 @@ func NewClientFromConfig(subscriptionID string, managementCert []byte, config Cl
...
+
+	err := client.importPublishSettingsFile(filePath)
+
+	return client, err
+}
+
+// NewClientFromPublishSettingsFileAndConfig creates a new Client using a publishsettings file and a given Client Config
+func NewClientFromPublishSettingsFileAndConfig(filePath string, config ClientConfig) (Client, error) {
+
+	client := Client{
```
Comment (ahmetalpbalkan) [inline management/client.go]: please use `NewClientFromConfig` to construct the client here. you're duplicating the logic.

File: management/client.go [hunk 4]
```diff
@@ -54,6 +54,30 @@ func NewClientFromConfig(subscriptionID string, managementCert []byte, config Cl
...
+	}
+
+	err := client.importPublishSettingsFile(filePath)
+
+	return client, err
+}
+
+// NewClientFromPublishSettingsFileAndConfig creates a new Client using a publishsettings file and a given Client Config
+func NewClientFromPublishSettingsFileAndConfig(filePath string, config ClientConfig) (Client, error) {
+
```
Comment (ahmetalpbalkan) [inline management/client.go]: redundant empty line @waitingkuo

File: management/client.go [hunk 5]
```diff
@@ -54,6 +54,30 @@ func NewClientFromConfig(subscriptionID string, managementCert []byte, config Cl
...
+	client := Client{
+		managementURL: config.ManagementURL,
+	}
+
+	err := client.importPublishSettingsFile(filePath)
+
```
Comment (ahmetalpbalkan) [inline management/client.go]: redundant empty line

**Comment by ahmetalpbalkan** [comment]
I am mistaken. `client.importPublishSettingsFil`e has a side effect (i.e. setting a global variable), this method should be removed, instead it should return a piece of information (e.g. cert/subs-id) which be embedded in client. 

Somebody must change that non-idempotent behavior, otherwise we're far from an ideal solution. I know it's a bit hard but current design of the library is just wrong. cc:@paulmey

**Comment by ahmetalpbalkan** [comment]
@waitingkuo we addressed this in #65. let us know if this solves your problem.

**Pull request closed without merging**